kotlin-runtime/ftl-runtime/src/main/kotlin/xyz/block/ftl/schemaextractor/ExtractSchemaRule.kt

```diff
@@ -0,0 +1,318 @@
+  )
+
+  override fun visitAnnotationEntry(annotationEntry: KtAnnotationEntry) {
+    if (
+      bindingContext.get(
+        BindingContext.ANNOTATION,
+        annotationEntry
+      )?.fqName?.asString() != xyz.block.ftl.Verb::class.qualifiedName
+    ) {
+      return
+    }
+
+    runCatching {
+      val extractor = SchemaExtractor(this.bindingContext, annotationEntry)
+      val moduleName = annotationEntry.containingKtFile.packageFqName.moduleName()
+      val moduleData = extractor.extract()
+      modules[moduleName]?.let { it.decls += moduleData.decls }
+        ?: run { modules[moduleName] = moduleData }
+    }.onFailure {
+      when (it) {
+        is IgnoredModuleException -> return
+        else -> throw it
+      }
+    }
+  }
+
+  override fun postVisit(root: KtFile) {
+    val outputDirectory = File(output).also { Path.of(it.absolutePath).createDirectories() }
+
+    modules.toModules().forEach {
+      val file = File(outputDirectory.absolutePath, OUTPUT_FILENAME)
+      file.createNewFile()
+      val os = FileOutputStream(file)
+      os.write(it.encode())
+      os.close()
+    }
+  }
+
+  companion object {
+    const val OUTPUT_FILENAME = "schema.pb"
+
+    private fun Map<String, ModuleData>.toModules(): List<Module> {
+      return this.map {
+        Module(name = it.key, decls = it.value.decls.sortedBy { it.data_ == null }, comments = it.value.comments)
+      }
+    }
+  }
+}
+
+class IgnoredModuleException : Exception()
+class SchemaExtractor(val bindingContext: BindingContext, annotation: KtAnnotationEntry) {
+  private val callMatcher: Regex
+  private val verb: KtNamedFunction
+  private val module: KtDeclaration
+  private val decls: MutableSet<Decl> = mutableSetOf()
+  fun extract(): ModuleData {
+    val requestType = requireNotNull(verb.valueParameters.last().typeReference?.resolveType()) {
+      "Could not resolve verb request type"
+    }
+    val responseType = requireNotNull(verb.createTypeBindingForReturnType(bindingContext)?.type) {
+      "Could not resolve verb response type"
+    }
+
+    val metadata = mutableListOf<Metadata>()
+    extractIngress()?.apply { metadata.add(Metadata(ingress = this)) }
+    extractCalls()?.apply { metadata.add(Metadata(calls = this)) }
+
+    val verb = Verb(
+      name = requireNotNull(verb.name) { "Verbs must be named" },
+      request = requestType.toSchemaType().dataRef,
+      response = responseType.toSchemaType().dataRef,
+      metadata = metadata,
+      comments = verb.comments(),
+    )
+
+    decls.addAll(
+      listOf(
+        Decl(verb = verb),
+        Decl(data_ = requestType.toSchemaData()),
+        Decl(data_ = responseType.toSchemaData())
+      )
+    )
+
+    return ModuleData(decls = decls, comments = module.comments())
+  }
+
+  private fun extractIngress(): MetadataIngress? {
+    return verb.annotationEntries.firstOrNull {
+      bindingContext.get(BindingContext.ANNOTATION, it)?.fqName?.asString() == Ingress::class.qualifiedName
```

Comment:
A bit unfortunate that we can't get a reference to the concrete annotation class like we can with KSP :(

Reply:
yeah, there were some operations that were a lot more ergonomic with KSP :( one thing that's nice about detekt though it's it's a thin shim around kotlin PSI, which I found a lot more google-able than the KSP frameworks/schemas. so maybe easier to build on until KSP resources have improved a bit

this is more verbose but should still be deterministic since we're comparing fully qualified class names